fe/fe-core/src/main/java/com/starrocks/statistic/columns/PredicateColumnsMgr.java

```diff
@@ -0,0 +1,226 @@
+        Predicate<ColumnUsage> pred = useCasePredicate.and(x -> predicate.test(x.getTableName()));
+        if (FeConstants.runningUnitTest) {
+            return id2columnUsage.values().stream().filter(pred).collect(Collectors.toList());
+        } else {
+            return getStorage().queryGlobalState(tableName, useCases).stream().filter(pred)
+                    .collect(Collectors.toList());
+        }
+    }
+
+    //==================================== Maintenance ============================================ //
+    public void persist() {
+        getStorage().persist(id2columnUsage.values());
+    }
+
+    public void vacuum() {
+        long ttlHour = Config.statistic_predicate_columns_ttl_hours;
+        LocalDateTime ttlTime = TimeUtils.getSystemNow().minusHours(ttlHour);
+        Predicate<ColumnUsage> outdated = x -> x.getLastUsed().isBefore(ttlTime);
+
+        long before = id2columnUsage.size();
+        if (id2columnUsage.values().removeIf(outdated)) {
+            long after = id2columnUsage.size();
+            LOG.info("removed {} objects from predicate columns because of ttl {}", before - after,
+                    Config.statistic_predicate_columns_ttl_hours);
+        }
+
+        // If the process crashed before vacuum the storage, the storage may be different from in-memory state,
+        // but it doesn't matter. Because we will remove them finally.
+        getStorage().vacuum(ttlTime);
+    }
+
+    public void restore() {
+        List<ColumnUsage> state = getStorage().restore();
+        for (ColumnUsage usage : ListUtils.emptyIfNull(state)) {
+            id2columnUsage.merge(usage, usage, ColumnUsage::merge);
+        }
+    }
+
+    private PredicateColumnsStorage getStorage() {
+        return PredicateColumnsStorage.getInstance();
+    }
+
+    @VisibleForTesting
+    public void reset() {
+        id2columnUsage.clear();
+    }
+
+    public void startDaemon() {
+        DaemonThread.getInstance().start();
+    }
+
+    static class DaemonThread extends FrontendDaemon {
+
+        private static final DaemonThread INSTANCE = new DaemonThread();
+
+        public DaemonThread() {
+            super("PredicateColumnsDaemonThread", Config.statistic_predicate_columns_persist_interval_sec * 1000L);
+        }
+
+        public static DaemonThread getInstance() {
+            return INSTANCE;
+        }
+
+        @Override
+        protected void runAfterCatalogReady() {
+            setInterval(Config.statistic_predicate_columns_persist_interval_sec * 1000L);
+
+            PredicateColumnsMgr mgr = PredicateColumnsMgr.getInstance();
+            PredicateColumnsStorage storage = PredicateColumnsStorage.getInstance();
+
+            if (!storage.isRestored()) {
+                mgr.restore();
+                storage.finishRestore();
+            } else {
+                mgr.vacuum();
+                mgr.persist();
```

Comment:
`id2columnUsage` may be persisted multi times since it's not incremental changes?

Reply:
correct. currently it's not incremental, but consider that number of columns in a system would not be unlimited, usually it's like thousands, hundred thousands. so choosing to persist all of them every a few minutes is still pretty cheap.

on the contrary, it's much complicated to implement a incremental + ttl storage, i would think it's a kind of over design.